File: src/java/foundgine-runtime/src/main/java/com/foundgine/runtime/controlplane/auditlog/AuditEvent.java
package com.foundgine.runtime.controlplane.auditlog;

import java.time.Instant;

/**
 * Port of {@code Foundgine.Runtime.ControlPlane.AuditLog.AuditEvent}, together
 * with the {@code AuditCategory} enum declared in the same C# file.
 *
 * <p>An immutable record of one governance step for one tool call. Events
 * carry a fingerprint rather than raw request/result payloads, the same
 * evidence-not-payload discipline used by
 * {@code com.foundgine.core.execution.ExecutionReceipt}, so the audit log
 * can be retained and shipped without duplicating domain data.
 *
 * <p><b>Porting decision:</b> {@code DateTimeOffset} is ported as
 * {@link Instant}, consistent with other timestamp fields elsewhere in this
 * port (e.g. {@code ApprovalRequest.ApprovalGrant.decidedAt}).
 */
public record AuditEvent(
        AuditCategory category,
        String toolName,
        String actor,
        String tenant,
        String fingerprint,
        String summary,
        Instant occurredAt) {

    public enum AuditCategory {
        RISK_SCORED,
        POLICY_EVALUATED,
        APPROVAL_REQUESTED,
        APPROVAL_DECIDED,
        ROUTED,
        DENIED,
    }

    public static AuditEvent create(
            AuditCategory category,
            String toolName,
            String actor,
            String tenant,
            String fingerprint,
            String summary) {
        return new AuditEvent(category, toolName, actor, tenant, fingerprint, summary, Instant.now());
    }
}

File: src/csharp/Foundgine.Runtime/ControlPlane/AuditLog/AuditEvent.cs
namespace Foundgine.Runtime.ControlPlane.AuditLog;

/// <summary>The stage of tool-call governance an audit event records.</summary>
public enum AuditCategory
{
    RiskScored,
    PolicyEvaluated,
    ApprovalRequested,
    ApprovalDecided,
    Routed,
    Denied,
}

/// <summary>
/// An immutable record of one governance step for one tool call. Events
/// carry a fingerprint rather than raw request/result payloads, the same
/// evidence-not-payload discipline used by
/// <see cref="Foundgine.Core.Execution.ExecutionReceipt"/>, so the audit
/// log can be retained and shipped without duplicating domain data.
/// </summary>
public sealed record AuditEvent(
    AuditCategory Category,
    string ToolName,
    string Actor,
    string? Tenant,
    string Fingerprint,
    string Summary,
    DateTimeOffset OccurredAt)
{
    public static AuditEvent Create(
        AuditCategory category,
        string toolName,
        string actor,
        string? tenant,
        string fingerprint,
        string summary) =>
        new(category, toolName, actor, tenant, fingerprint, summary, DateTimeOffset.UtcNow);
}
